### PR TITLE
[BEAM-5320] Finish Python 3 porting for testing module

### PR DESCRIPTION
**Pull request by RobbeSneyders**
This is is part of a series of PRs with goal to make Apache Beam PY3 compatible. The proposal with the outlined approach has been documented here: https://s.apache.org/beam-python-3

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by RobbeSneyders** [comment]
R: @tvalentyn @splovyt @markflyhigh

**Comment by markflyhigh** [comment]
LGTM. Thanks!

File: sdks/python/apache_beam/testing/util.py [hunk 1]
```diff
@@ -113,14 +113,19 @@ def matcher(elements):
...
-  expected = list(expected)
 
   def _equal(actual):
-    sorted_expected = sorted(expected)
-    sorted_actual = sorted(actual)
-    if sorted_expected != sorted_actual:
+    expected_list = list(expected)
+    for element in actual:
+      try:
+        expected_list.remove(element)
```
Comment (aaltay) [inline sdks/python/apache_beam/testing/util.py]: This no longer does what the above comment about sorting. Please update the comment. (or do it in a follow up)

File: sdks/python/apache_beam/transforms/util.py [hunk 2]
```diff
@@ -223,7 +223,7 @@ def __init__(self,
     if target_batch_duration_secs and target_batch_duration_secs <= 0:
       raise ValueError("target_batch_duration_secs (%s) must be positive" % (
           target_batch_duration_secs))
-    if max(0, target_batch_overhead, target_batch_duration_secs) == 0:
+    if not (target_batch_overhead or target_batch_duration_secs):
```
Comment (aaltay) [inline sdks/python/apache_beam/transforms/util.py]: What is the reason for this change?

I do not think this is checking for at least one number being > 0. For example not (-1 or -3) will not raise this error.

Comment (RobbeSneyders) [reply sdks/python/apache_beam/transforms/util.py]: These are changes introduced in #7022. This PR is rebased on top of that one, since those changes have an impact on the testing module.

File: sdks/python/apache_beam/testing/util.py (hunk 1 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/testing/util.py]: Updated in #7022. I'll rebase this one.

File: sdks/python/apache_beam/transforms/util.py (hunk 2 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/transforms/util.py]: Previous if checks make sure those numbers are non-negative.

**Comment by RobbeSneyders** [comment]
I just noticed in the logs that there was an error running the py27-gcp tests. However the Precommit says it has been successful. Is this a bug in the pipeline?

**Comment by markflyhigh** [comment]
https://issues.apache.org/jira/browse/BEAM-6048 is tracking this issue. https://github.com/apache/beam/pull/7048 should fix it. Please rebase and test again.

File: sdks/python/apache_beam/testing/test_utils.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def compute_hash(content, hashing_alg=DEFAULT_HASHING_ALG):
   content.sort()
   m = hashlib.new(hashing_alg)
   for elem in content:
-    m.update(str(elem).encode('utf-8'))
+    m.update(elem)
```
Comment (RobbeSneyders) [inline sdks/python/apache_beam/testing/test_utils.py]: This fails on `None` values introduced in [this test](https://github.com/apache/beam/blob/8db6b63276f6e749f095b5817b22efb431e22e6b/sdks/python/apache_beam/io/gcp/tests/bigquery_matcher_test.py#L54).

However the documentation of the current function indicates that it expects a list of strings. I also can't think of a scenario where BigQuery returns a list of `None` values for rows from a query. I'm therefore wondering if this is a problem with the current function, or the test should be changed.

Previously, this function just casted `None` to `"None"` before calculating the hash.

Comment (RobbeSneyders) [reply sdks/python/apache_beam/testing/test_utils.py]: PTAL @tvalentyn @markflyhigh

Comment (RobbeSneyders) [reply sdks/python/apache_beam/testing/test_utils.py]: On Python 3, `str(None).encode('utf-8')` results in `'b"None"'`. Checking for `bytes` fixes this.
I also moved this conversion to the start of the method, since `sort()` otherwise fails with a `TypeError` on Python 3 because of unorderable types.

**Comment by markflyhigh** [comment]
LGTM
Thanks

**Comment by swegner** [comment]
FYI, it looks like this PR may have broken Python post-commit: https://builds.apache.org/job/beam_PostCommit_Python_Verify/6617/

**Comment by aaltay** [comment]
Relevant error:

15:24:55 ======================================================================
15:24:55 FAIL: test_bigquery_tornadoes_it (apache_beam.examples.cookbook.bigquery_tornadoes_it_test.BigqueryTornadoesIT)
15:24:55 ----------------------------------------------------------------------
15:24:55 Traceback (most recent call last):
15:24:55   File "/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Python_Verify/src/sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes_it_test.py", line 73, in test_bigquery_tornadoes_it
15:24:55     test_pipeline.get_full_options_as_args(**extra_opts))
15:24:55   File "/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Python_Verify/src/sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes.py", line 90, in run
15:24:55     write_disposition=beam.io.BigQueryDisposition.WRITE_TRUNCATE)
15:24:55   File "/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Python_Verify/src/sdks/python/apache_beam/pipeline.py", line 425, in __exit__
15:24:55     self.run().wait_until_finish()
15:24:55   File "/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Python_Verify/src/sdks/python/apache_beam/pipeline.py", line 418, in run
15:24:55     return self.runner.run_pipeline(self)
15:24:55   File "/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Python_Verify/src/sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py", line 67, in run_pipeline
15:24:55     hc_assert_that(self.result, pickler.loads(on_success_matcher))
15:24:55 AssertionError: 
15:24:55 Expected: (Test pipeline expected terminated in state: DONE and Expected checksum is 83789a7c1bca7959dcf23d3bc37e9204e594330f)
15:24:55      but: Expected checksum is 83789a7c1bca7959dcf23d3bc37e9204e594330f Actual checksum is d860e636050c559a16a791aff40d6ad809d4daf0
15:24:55

**Comment by aaltay** [comment]
This is probably related to how the hash is computed rather than an actual change in data. I will suggest:
1. Reverting to get tests back to green.
2. Looking at what in the failing tests is not resulting in a different hash result.
3. After (2) adjusting either the hash function or the expected output of the test.

My guess is (2) is hashing some non-string values which has a different hashing behavior now.

**Comment by tvalentyn** [comment]
I agree with both assessments.

**Comment by RobbeSneyders** [comment]
What is the best way to check this test? I see it is not being run in the precommit test suite.

**Comment by aaltay** [comment]
It runs as part of post commit tests. You can add a comment "Run Python PostCommit" to trigger it.

**Comment by aaltay** [comment]
https://github.com/apache/beam/pull/7112 -> cherry picking revert to release branch. After this I will assign bug back to you @RobbeSneyders for the fix and roll forward of this PR again.
